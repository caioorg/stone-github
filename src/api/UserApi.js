import instance from './index'

class User {
  static async getUser(user) {
    let status = {}
    await instance
      .get(`/users/${user}`)
      .then(e => {
        if (e.status === 200) status = { error: false, data: e.data }
      })
      .catch(err => {
        status = { error: true, message: err }
      })

    return status
  }

  static async getBibliographyUser(user) {
    let status = {}

    await instance.get(`/users/${user}`).then(e => {
      if (e.status === 200) {
        status = {
          error: false,
          data: {
            id: e.data.id,
            login: e.data.login,
            avatar_url: e.data.avatar_url,
            html_url: e.data.html_url,
            name: e.data.name,
            company: e.data.company,
            location: e.data.location,
            bio: e.data.bio,
            public_repos: e.data.public_repos,
            followers: e.data.followers,
            email: e.data.email,
            following: e.data.following
          }
        }
      }
    })

    return status
  }

  static async getUserRepositories(user) {
    let status = {}
    await instance
      .get(`/users/${user}/repos`)
      .then(e => {
        if (e.status === 200) status = { error: false, data: e.data }
      })
      .catch(err => {
        status = { error: true, message: err }
      })
    return status
  }
}

export default User
