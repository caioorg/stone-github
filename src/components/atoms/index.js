export { default as Icon } from './Icon'
export { default as Input } from './Input'
export { default as Label } from './Label'
export { default as Loader } from './Loader'
export { default as Row } from './Row'
