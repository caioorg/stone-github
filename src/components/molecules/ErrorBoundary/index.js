import React, { Component } from 'react'

class ErrorBoundary extends Component {
  constructor(props) {
    super(props)
    this.state = {
      error: false
    }
  }

  componentDidCatch(error) {
    this.setState({ error: true })
  }

  render() {
    const { error } = this.state
    return error ? <div>Aplicação Indisponivel</div> : this.props.children
  }
}

export default ErrorBoundary
