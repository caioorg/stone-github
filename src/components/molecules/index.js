export { default as ErrorBoundary } from './ErrorBoundary'
export { default as SearchUser } from './SearchUser'
export { default as Bibliography } from './Bibliography'
export { default as ListOfRepositories } from './ListOfRepositories'
