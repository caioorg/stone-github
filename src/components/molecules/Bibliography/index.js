import React, { useEffect, useState } from 'react'
import { withRouter, Link } from 'react-router-dom'
import User from '@/api/UserApi'
import { Label, Row, Icon } from '@/components/atoms'

import { Wrapper, Container, Image } from './styled'

const Bibliography = ({ ...props }) => {
  const [bibliography, setBibliography] = useState({})
  useEffect(() => {
    User.getBibliographyUser(props.match.params.user).then(e => {
      if (!e.error) setBibliography(e.data)
    })
  }, [props.match.params.user])

  return (
    <Wrapper>
      <Link to="/">
        <Icon id="arrow-left" small />
        Pesquisar Novamente
      </Link>
      <Container>
        {bibliography.avatar_url && (
          <Row>
            <a href={bibliography.html_url} target="_blank" without rel="noopener noreferrer">
              <Image src={bibliography.avatar_url} alt={bibliography.name} />
            </a>
          </Row>
        )}

        {bibliography.name && (
          <Row>
            <Label size={12} weight="bold" color="#000000">
              Name:
            </Label>
            <Label bottom={5} size={14}>
              {bibliography.name}
            </Label>
          </Row>
        )}

        {bibliography.id && (
          <Row>
            <Label bottom={5} size={12} weight="bold" color="#000000">
              Identifier:
            </Label>
            <Label bottom={10} size={14}>
              {bibliography.id}
            </Label>
          </Row>
        )}

        {bibliography.login && (
          <Row>
            <Label bottom={5} size={12} weight="bold" color="#000000">
              Username:
            </Label>
            <Label bottom={10} size={14}>
              {bibliography.login}
            </Label>
          </Row>
        )}

        {bibliography.email && (
          <Row>
            <Label bottom={5} size={12} weight="bold" color="#000000">
              Email:
            </Label>
            <Label bottom={10} size={14}>
              {bibliography.email}
            </Label>
          </Row>
        )}

        {bibliography.location && (
          <Row>
            <Label bottom={5} size={12} weight="bold" color="#000000">
              Location:
            </Label>
            <Label bottom={10} size={14}>
              {bibliography.location}
            </Label>
          </Row>
        )}

        {bibliography.company && (
          <Row>
            <Label bottom={5} size={12} weight="bold" color="#000000">
              Company Current:
            </Label>
            <Label bottom={5} size={14}>
              {bibliography.company}
            </Label>
          </Row>
        )}

        {bibliography.bio && (
          <Row>
            <Label bottom={5} size={12} weight="bold" color="#000000">
              Bibliography:
            </Label>
            <Label bottom={5} size={14}>
              {bibliography.bio}
            </Label>
          </Row>
        )}

        {bibliography.followers !== null && (
          <Row>
            <Label bottom={5} size={12} weight="bold" color="#000000">
              Followers:
            </Label>
            <Label bottom={5} size={14}>
              {bibliography.followers}
            </Label>
          </Row>
        )}

        {bibliography.following !== null && (
          <Row>
            <Label bottom={5} size={12} weight="bold" color="#000000">
              Following:
            </Label>
            <Label bottom={5} size={14}>
              {bibliography.following}
            </Label>
          </Row>
        )}
      </Container>
    </Wrapper>
  )
}

export default withRouter(Bibliography)
