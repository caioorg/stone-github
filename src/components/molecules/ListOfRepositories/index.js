import React, { useEffect, useState } from 'react'
import { withRouter } from 'react-router-dom'
import User from '@/api/UserApi'
import { Icon, Label, Row } from '@/components/atoms'
import { Container, Card, Header, HeaderIndicators, HeaderTitle, Content } from './styled'

const ListOfRepositories = ({ ...props }) => {
  const [repo, setRepo] = useState([])
  useEffect(() => {
    User.getUserRepositories(props.match.params.user).then(e => {
      const result = e.data.sort((a, b) => {
        if (a.stargazers_count > b.stargazers_count) return -1
        if (a.stargazers_count < b.stargazers_count) return 1
        return 0
      })
      setRepo(result)
    })
  }, [props.match.params.user])

  return (
    <Container>
      {repo &&
        repo.map(
          ({
            language,
            forks_count,
            stargazers_count,
            watchers_count,
            svn_url,
            name,
            full_name,
            id,
            description
          }) => {
            return (
              <Card href={svn_url} key={id} target="_blank">
                <Header>
                  <HeaderTitle>
                    <Label weight="bold" size={16} display="inline-block">
                      {name && name.replace(/-/g, ' ')}
                    </Label>
                  </HeaderTitle>
                  <HeaderIndicators>
                    <Label display="inline-block">
                      <Icon id="star" small /> ({stargazers_count && stargazers_count})
                    </Label>
                    <Label display="inline-block">
                      <Icon id="eye" small /> ({watchers_count && watchers_count})
                    </Label>
                    <Label display="inline-block">
                      <Icon id="fork" small /> ({forks_count && forks_count})
                    </Label>
                  </HeaderIndicators>
                </Header>
                <Content>
                  <Row>
                    <Label bottom={5} size={12} weight="bold" color="#000000">
                      Full Name Repository:
                    </Label>
                    <Label bottom={10} size={14}>
                      {full_name && full_name}
                    </Label>
                  </Row>
                  <Row>
                    <Label bottom={5} size={12} weight="bold" color="#000000">
                      Language:
                    </Label>
                    <Label bottom={10} size={14}>
                      {language && language}
                    </Label>
                  </Row>
                  <Row>
                    <Label bottom={5} size={12} weight="bold" color="#000000">
                      Description:
                    </Label>
                    <Label bottom={10} size={14}>
                      {description && description}
                    </Label>
                  </Row>
                </Content>
              </Card>
            )
          }
        )}
    </Container>
  )
}

export default withRouter(ListOfRepositories)
