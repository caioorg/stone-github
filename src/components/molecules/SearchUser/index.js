import React, { useState, Fragment } from 'react'
import { withRouter } from 'react-router-dom'
import { notify } from 'react-notify-toast'
import { Input, Icon, Loader } from '@/components/atoms'
import { appMessages } from '@/helpers'
import User from '@/api/UserApi'
import { Container, FormWrapper, Button } from './styled'

export const SearchUser = ({ ...props }) => {
  const [search, setSearch] = useState('')
  const [isLoading, setIsLoading] = useState(false)

  const renderNotify = message => {
    return notify.show(message, 'custom', 5000, {
      background: '#bb2939',
      text: '#FFFFFF'
    })
  }

  const onHandleChange = e => {
    e.preventDefault()
    setIsLoading(true)

    if (search === '') {
      setIsLoading(false)
      return renderNotify(appMessages.search['user-empty'])
    }

    User.getUser(search)
      .then(e => {
        if (e.error || (e.message && e.message.request.status === 404)) {
          setIsLoading(false)
          return renderNotify(appMessages.search['user-not-found'])
        }
        if (!e.error && Object.keys(e.data).length > 0) {
          setIsLoading(false)
          return props.history.push(`/user/${e.data.login}/repository`)
        }
      })
      .catch(e => {
        setIsLoading(false)
        return renderNotify(appMessages.system.error)
      })
      .finally(() => {
        setIsLoading(false)
      })
  }

  return (
    <Fragment>
      {isLoading && <Loader />}
      <Container>
        <FormWrapper onSubmit={e => onHandleChange(e)}>
          <Input onChange={e => setSearch(e)} placeholder={appMessages.search.placeholder} />
          <Button title="Search User" onClick={e => onHandleChange(e)}>
            <Icon id="search" small />
          </Button>
        </FormWrapper>
      </Container>
    </Fragment>
  )
}

export default withRouter(SearchUser)
