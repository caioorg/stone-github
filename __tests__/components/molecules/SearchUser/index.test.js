/* eslint-disable no-undef */
import React from 'react'
import { SearchUser } from '@/components/molecules'
import { MemoryRouter } from 'react-router'
import { getUser } from '@/api/UserApi'

jest.mock('@/api/UserApi', () => {
  return {
    getUser: jest.fn(async () => ({
      error: null,
      data: {}
    }))
  }
})

beforeEach(() => {
  getUser.mockClear()
})

describe('Testing SearchUser Component', () => {
  it('Expect to render correctly ', async () => {
    const { asFragment } = await global.render(
      <MemoryRouter>
        <SearchUser />
      </MemoryRouter>
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it('Expect user to find the person', async () => {
    getUser.mockImplementation(async () => ({
      error: false,
      data: {
        login: 'caio'
      }
    }))

    const { container } = await global.render(
      <MemoryRouter>
        <SearchUser />
      </MemoryRouter>
    )

    global.fireEvent.change(container.querySelector('input'), { target: { value: 'caioorg' } })
    global.fireEvent.click(container.querySelector('button'))
    expect(getUser).toHaveBeenCalledTimes(1)
  })
})
