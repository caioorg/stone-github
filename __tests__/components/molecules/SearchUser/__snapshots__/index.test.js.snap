// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing SearchUser Component Expect to render correctly  1`] = `
<DocumentFragment>
  .c5 {
  display: inline-block;
  vertical-align: middle;
  width: 20px;
  height: 20px;
}

.c2 {
  display: inline-block;
  width: 100%;
  position: relative;
  border-bottom: 2px solid #34303f;
}

.c3 {
  width: 100%;
  padding: 20px 40px;
  background: #ffffff;
  border-radius: 10px;
  border: 0;
  font-size: 20px;
  color: #5e5e6a;
  font-family: 'PT Sans Narrow',sans-serif;
}

.c3::-webkit-input-placeholder {
  color: #5e5e6a61;
  opacity: 1;
  font-weight: bold;
}

.c3::-moz-placeholder {
  color: #5e5e6a61;
  opacity: 1;
  font-weight: bold;
}

.c3:-ms-input-placeholder {
  color: #5e5e6a61;
  opacity: 1;
  font-weight: bold;
}

.c3::placeholder {
  color: #5e5e6a61;
  opacity: 1;
  font-weight: bold;
}

.c3:focus {
  border: 0;
}

.c0 {
  width: 40%;
  margin: 0 auto;
  position: relative;
  top: 50%;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
}

.c1 {
  position: relative;
  margin-bottom: 20px;
}

.c4 {
  width: 60px;
  height: 40px;
  border-radius: 50px;
  border: 0;
  background-color: #e02e41;
  position: absolute;
  right: 40px;
  top: 50%;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  cursor: pointer;
}

.c4:hover {
  background: #bb2939;
}

.c4 svg {
  fill: #fff;
}

<div
    class="c0"
  >
    <form
      class="c1"
    >
      <div
        class="c2"
      >
        <input
          class="c3"
          placeholder="What is your GitHub user?"
        />
      </div>
      <button
        class="c4"
        title="Search User"
      >
        <svg
          class="c5 icon icon-search"
        >
          <use
            xlink:href="test-file-stub#search"
          />
        </svg>
      </button>
    </form>
  </div>
</DocumentFragment>
`;
